lib/PAUSE/mldistwatch.pm

```diff
@@ -347,31 +349,31 @@ sub _do_the_database_work {
     # Either we're doing Perl 6...
     if ($dio->perl_major_version == 6) {
       if ($dio->p6_dist_meta_ok) {
-        if (my $err = $dio->p6_index_dist) {
-          $dio->alert($err);
+        if (my $err = $dio->p6_index_dist($ctx)) {
+          $ctx->alert($err);
           $dbh->rollback;
         } else {
           $dbh->commit;
         }
       }
       else {
-        $dio->alert("Meta information of Perl 6 dist is invalid");
+        $ctx->alert("Meta information of Perl 6 dist is invalid");
         $dbh->rollback;
       }
 
       return 1;
     }
 
     # ...or else Perl 5...
-    $dio->examine_pms;      # will switch user
+    $dio->examine_pms($ctx);      # will switch user
 
     my $main_pkg = $dio->_package_governing_permission;
 
     if ($self->permissions->userid_has_permissions_on_package($dio->{USERID}, $main_pkg)) {
       $dbh->commit;
     } else {
-      $dio->alert("Uploading user has no permissions on package $main_pkg");
-      $dio->{NO_DISTNAME_PERMISSION} = 1;
+      $ctx->alert("Uploading user has no permissions on package $main_pkg");
+      $ctx->add_dist_error(ERROR('no_distname_permission'));
```

Comment:
Maybe all add_dist_error calls should actually be aborts?

Reply:
These two can't (yet) because they happen outside the error handler.